clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -272,55 +280,96 @@ private boolean isMuted(TopicPartition tp, long now) {
+                        // statement here to log info of the batches that violate the invariant.
+                        if (batch.finalState() == null) {
+                            batch.abortRecordAppends();
                             expiredBatches.add(batch);
-                            batchIterator.remove();
                         } else {
-                            // Stop at the first batch that has not expired.
-                            break;
+                            log.warn("{} batch crated at {} gets unexpected final state : {}",
+                                batch.topicPartition, batch.createdMs, batch.finalState());
                         }
+                    } else {
+                        maybeUpdateNextBatchExpiryTime(batch);
+                        break;
+                    }
+                }
+                if (partitionInFlightBatches.isEmpty())
+                    inFlightBatches.remove(topicPartition);
+            }
+            Deque<ProducerBatch> deque = entry.getValue();
+            synchronized (deque) {
+                while (!deque.isEmpty()) {
+                    ProducerBatch batch = deque.getFirst();
+                    if (batch.hasReachedDeliveryTimeout(deliveryTimeoutMs, now)) {
+                        deque.poll();
+                        batch.abortRecordAppends();
+                        expiredBatches.add(batch);
+                    } else {
+                        maybeUpdateNextBatchExpiryTime(batch);
+                        break;
                     }
                 }
             }
         }
         return expiredBatches;
     }
 
+    public List<ProducerBatch> inFlightBatches(TopicPartition tp) {
+        return inFlightBatches.containsKey(tp) ? inFlightBatches.get(tp) : new ArrayList<>();
+    }
+
     /**
-     * Re-enqueue the given record batch in the accumulator to retry
+     * Re-enqueue the given record batch in the accumulator to retry if the batch has not reached deliveryTimeoutMs.
+     * Remove the batch from the list of inflight batches.
      */
     public void reenqueue(ProducerBatch batch, long now) {
-        batch.reenqueued(now);
-        Deque<ProducerBatch> deque = getOrCreateDeque(batch.topicPartition);
-        synchronized (deque) {
-            if (transactionManager != null)
-                insertInSequenceOrder(deque, batch);
-            else
-                deque.addFirst(batch);
+        if (!batch.hasReachedDeliveryTimeout(deliveryTimeoutMs, now)) {
+            batch.reenqueued(now);
+            Deque<ProducerBatch> deque = getOrCreateDeque(batch.topicPartition);
+            synchronized (deque) {
+                if (transactionManager != null)
+                    insertInSequenceOrder(deque, batch);
+                else
+                    deque.addFirst(batch);
+            }
+        } else {
+            batch.abortRecordAppends();
```

Comment:
I think the only issue with reenqueing unconditionally is that the batch will then be drained even if it is expired, since `accumulator.drain` is called before `accumulator.expiredBatches` in the background thread. This could violate the contract.

That said, we need to complete the batch and deallocate it over here, otherwise it seems to be dropped on the flor.

Reply:
updated to check whether a batch has reached `deliveryTimeoutMs` or not in `Sender.canRetry`.  With this change, we do not need to check whether a batch has timeout or not in `RecordAccumuator.reenqueue`. 

       private boolean canRetry(ProducerBatch batch, ProduceResponse.PartitionResponse response, long now) {
            return !batch.hasReachedDeliveryTimeout(accumulator.getDeliveryTimeoutMs(), now) &&